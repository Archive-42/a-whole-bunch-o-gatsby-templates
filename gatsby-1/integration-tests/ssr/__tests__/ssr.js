const fetch = require(`node-fetch`)
const execa = require(`execa`)
const fs = require(`fs-extra`)
const path = require(`path`)

function fetchUntil(url, filter, timeout = 1000) {
  return new Promise(resolve => {
    fetch(url).then(res => {
      if (filter(res)) {
        resolve(res)
      } else {
        setTimeout(() => {
          resolve(fetchUntil(url, filter, timeout))
        }, timeout)
      }
    })
  })
}

describe(`SSR`, () => {
  test(`is run for a page when it is requested`, async () => {
    const html = await fetch(`http://localhost:8000/`).then(res => res.text())

    expect(html).toMatchSnapshot()
  })

  test(`dev & build outputs match`, async () => {
    const childProcess = await execa(`yarn`, [`test-output`])

    expect(childProcess.code).toEqual(0)

    // Additional sanity-check
    expect(String(childProcess.stdout)).toContain(
      `testing these paths for differences between dev & prod outputs`
    )
  }, 15000)

  test(`it generates an error page correctly`, async () => {
    const src = path.join(__dirname, `/fixtures/bad-page.js`)
    const dest = path.join(__dirname, `../src/pages/bad-page.js`)
    fs.copySync(src, dest)

    const pageUrl = `http://localhost:8000/bad-page/`
    // Poll until the new page is bundled (so starts returning a non-404 status).
    const rawDevHtml = await fetchUntil(pageUrl, res => {
      return res
    }).then(res => res.text())

    expect(rawDevHtml).toMatch("<h1>Failed to Server Render (SSR)</h1>")
    expect(rawDevHtml).toMatch("<h2>Error message:</h2>")
    expect(rawDevHtml).toMatch("<p>window is not defined</p>")
    // html should contain stacktrace to bad-page
    expect(rawDevHtml).toMatch(/at Component \(.+?(?=bad-page.js)[^)]+\)/)
    await fs.remove(dest)

    // After the page is gone, it'll 404.
    // TODO FIX as this isn't working
    // await fetchUntil(pageUrl, res => res.status === 404)
  }, 15000)
})
