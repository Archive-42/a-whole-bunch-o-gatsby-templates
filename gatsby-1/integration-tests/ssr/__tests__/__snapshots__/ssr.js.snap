// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SSR is run for a page when it is requested 1`] = `"<!DOCTYPE html><html><head><meta charSet=\\"utf-8\\"/><meta http-equiv=\\"x-ua-compatible\\" content=\\"ie=edge\\"/><meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1, shrink-to-fit=no\\"/><link data-identity=\\"gatsby-dev-css\\" rel=\\"stylesheet\\" type=\\"text/css\\" href=\\"/commons.css\\"/><meta name=\\"note\\" content=\\"environment=development\\"/><script src=\\"/socket.io/socket.io.js\\"></script></head><body><div id=\\"___gatsby\\"><div style=\\"outline:none\\" tabindex=\\"-1\\" id=\\"gatsby-focus-wrapper\\"><div><h1 class=\\"hi\\">Hello world</h1></div></div><div id=\\"gatsby-announcer\\" style=\\"position:absolute;top:0;width:1px;height:1px;padding:0;overflow:hidden;clip:rect(0, 0, 0, 0);white-space:nowrap;border:0\\" aria-live=\\"assertive\\" aria-atomic=\\"true\\"></div></div><script src=\\"/polyfill.js\\" nomodule=\\"\\"></script><script src=\\"/framework.js\\"></script><script src=\\"/commons.js\\"></script></body></html>"`;
